### PR TITLE
feat: Azbatch executor

### PR DESCRIPTION
**Pull request by jakevc**
### Description

Finish remaining issues with the original implementation of the Azure Batch Executor originally tackled by @andreas-wilm in #533. 

- simplified command line options
- batch configuration as a class with sensible defaults, or environment variable overrides.
- made use of existing AzBlob helper where possible.
- updates dockerfile with necessary deps
- container jakevc/snakemake:latest contains these changes for testing

Toyed with the idea of having more configuration from command line options, but this litters the codebase with redundant code (each command line option needs to be repeated a few times in __init__.py, scheduler.py, workflow.py, azure_batch.py). Env Variables seem appropriate with sensible defaults. 

### QC

* [x] The PR contains a test case for the changes or the changes are already covered by an existing test case.
* [x] The documentation (`docs/`) is updated to reflect the changes or this is not necessary (e.g. if the change does neither modify the language nor the behavior or functionalities of Snakemake).

### Example

Example invocation:

```bash
export AZ_BLOB_ACCOUNT_URL=*********
export AZ_BATCH_ACCOUNT_KEY=*********

snakemake \
    --jobs 3 \
    -rpf --verbose --default-remote-prefix snaketest \
    --use-conda \
    --default-remote-provider AzBlob \
    --envvars AZ_BLOB_ACCOUNT_URL \
    --az-batch \
    --container-image jakevc/snakemake \
    --az-batch-account-url **********
```

### CODE REVIEW

**Comment by vsmalladi** [comment]
Is this PR ready to test?

**Comment by jakevc** [comment]
yes please give it a try and let me know if you have any issues/recommendations

File: snakemake/executors/azure_batch.py
Comment (johanneskoester) [inline snakemake/executors/azure_batch.py:22]: Please move this import into the method where it is used. Otherwise, AzBlob will be loaded even if somebody uses nothing from azure, essentially turning the required libs a necessary dependency.

File: snakemake/__init__.py
Comment (johanneskoester) [inline snakemake/__init__.py:2805]: What is the rationale for representing this as url instead of expecting a location and an account flag?

**Review by johanneskoester — CHANGES_REQUESTED** [review]
Awesome work! Very clean code! Just a few comments below.

And then: I've contacted MS about credits, hopefully we can add a testcase before merging.

**Comment by johanneskoester** [comment]
Hey Jake. As Adreas Wilm is already an author of the Snakemake paper because of his contribution to Azure Batch support, it makes total sense to make you a co-author as well in the upcoming revision. Do you agree? If so, please give me your affiliation here.

File: snakemake/__init__.py
Comment (jakevc) [reply snakemake/__init__.py:2805]: Since the full batch_account_url is used as an argument to BatchServiceClient in line 165, and the batch_account_name is used for the SharedKeyCredentials in line 162, but the location is never used. 

It's a tradeoff I guess. Having it the other way would mean increasing the probability of user error when constructing the batch_account_url, if either the location or the name is mis-specified. This way someone only has to enter the correct value once. Also since location is not used, that flag would be an additional 5ish lines of code throughout the argument parsing logic that really is only used to construct the full batch_account_url and nothing else..

**Comment by jakevc** [comment]
> Hey Jake. As Adreas Wilm is already an author of the Snakemake paper because of his contribution to Azure Batch support, it makes total sense to make you a co-author as well in the upcoming revision. Do you agree? If so, please give me your affiliation here.

Yes that sounds good, thank you!

Please include these two affiliations:
- Earle A. Chiles Research Institute, Portland, Oregon.
- Providence Cancer Institute, Portland, Oregon.

File: docs/executor_tutorial/azure_batch.rst
Comment (vsmalladi) [inline docs/executor_tutorial/azure_batch.rst:81]: Should change variable for storage accoutn name to `stgacct` from `storageacct`
```
stgkey=$(az storage account keys list -g $resgroup -n $stgacct | head -n1 | cut -f 3)
```

File: docs/executor_tutorial/azure_batch.rst
Comment (vsmalladi) [inline docs/executor_tutorial/azure_batch.rst:119]: Stoge batch account keys

```
az_batch_account_key=$(az batch account keys list --resource-group $resgroup --name $accountname -o tsv | head -n1 | cut -f2)
```

File: docs/executor_tutorial/azure_batch.rst
Comment (vsmalladi) [inline docs/executor_tutorial/azure_batch.rst:127]: Do you want users to generate this in CLI or by Portal?

File: docs/executor_tutorial/azure_batch.rst
Comment (vsmalladi) [inline docs/executor_tutorial/azure_batch.rst:136]: Where is this file, link doesn't work

File: docs/executor_tutorial/azure_batch.rst
Comment (jakevc) [reply docs/executor_tutorial/azure_batch.rst:127]: CLI would be nice but I can't seem to find the command to generate an SAS url for a storage account that produces the full URI.

**Review by vsmalladi — COMMENTED** [review]
When I ran through the turorial I got the following error:

```
/Users/venkatmalladi/opt/anaconda3/envs/snakemake/lib/python3.11/site-packages/snakemake/executors/azure_batch.py:352: RuntimeWarning: coroutine 'sleep' was never awaited
  sleep()
```

File: docs/executor_tutorial/azure_batch.rst
Comment (vsmalladi) [reply docs/executor_tutorial/azure_batch.rst]: Should we include this in the repo?

**Comment by vsmalladi** [comment]
Overall this is great. Does this scale to be a new node for each task or all the tasks running on a single node?

Just curious since conda is still used vs docker

**Comment by vsmalladi** [comment]
Got this error on mac os: 

```
The Python 3 package 'azure-storage-blob' need to be installed to use Azure Storage remote() file functionality. No module named 'azure'
```

Verified that azure-storage-blob was installed using pip. Any ideas?

**Comment by jakevc** [comment]
> When I ran through the tutorial I got the following error:
> 
> ```
> /Users/venkatmalladi/opt/anaconda3/envs/snakemake/lib/python3.11/site-packages/snakemake/executors/azure_batch.py:352: RuntimeWarning: coroutine 'sleep' was never awaited
>   sleep()
> ```
> 
> No issue once i exported AZ_BATCH_ACCOUNT_KEY

fixed this

**Comment by jakevc** [comment]
> Overall this is great. Does this scale to be a new node for each task or all the tasks running on a single node?
> 
> Just curious since conda is still used vs docker

Currently the pool autoscaling is not enabled, so it will distribute tasks across one node (however big it is). In the future I would like to include autoscaling, but need to do some testing to configure the most appropriate autoscaling formula, discussed here: https://learn.microsoft.com/en-us/azure/batch/batch-automatic-scaling

File: docs/executor_tutorial/azure_batch.rst
Comment (vsmalladi) [reply docs/executor_tutorial/azure_batch.rst:125]: export sastoken=$(az storage account generate-sas \
    --account-name $stgacct \
    --permissions acdlrw \
    --services bf \
    --resource-types co \
    --expiry $expiry_date)

export sas_url=$(az storage account show-connection-string \
    -g $resgroup \
    -n $stgacct \
    --protocol https \
    -o tsv | cut -f5,9 -d ';' | cut -f 2 -d '=')

 export storage_account_url_with_sas='${sas_url}${sastoken}'

Something like this. can be made better.

**Comment by vsmalladi** [comment]
@jakevc I tried to run this again.

Based on the tutorial the base snakemake image might not have the azure plugins cause I get this error:

`The Python 3 package 'azure-storage-blob' need to be installed to use Azure Storage remote() file functionality. No module named 'azure'`

Thoughts?

**Comment by jakevc** [comment]
> @jakevc I tried to run this again.
> 
> Based on the tutorial the base snakemake image might not have the azure plugins cause I get this error:
> 
> `The Python 3 package 'azure-storage-blob' need to be installed to use Azure Storage remote() file functionality. No module named 'azure'`
> 
> Thoughts?

I will try this out in a fresh environment and see if I can reproduce

**Comment by jakevc** [comment]
> scale to be a new node for each task or all the tasks running on a single node?
> 
> Just curious since conda is still used vs docker

Was able to implement basic autoscaling, and enabled some other configuration parameters that enable task distribution on dedicated ( non autoscaling ) nodes as well.

**Comment by vsmalladi** [comment]
> > scale to be a new node for each task or all the tasks running on a single node?
> > Just curious since conda is still used vs docker
> 
> Was able to implement basic autoscaling, and enabled some other configuration parameters that enable task distribution on dedicated ( non autoscaling ) nodes as well.

Oh that is cool. Looking forward to testing that out.

**Comment by jakevc** [comment]
> @jakevc I tried to run this again.
> 
> Based on the tutorial the base snakemake image might not have the azure plugins cause I get this error:
> 
> `The Python 3 package 'azure-storage-blob' need to be installed to use Azure Storage remote() file functionality. No module named 'azure'`
> 
> Thoughts?

Okay the method I am using to create an environment and test this branch is the following, and you should be able to reproduce if you follow these instructions.

Fist checkout my fork branch `azbatch-executor` and using either mamba or conda to create the testing environment. Mamba solves the environment a lot faster.  From the root of the snakemake repo:

```
mamba env create -f test-environment.yml -n snakemake-testing
conda activate snakemake-testing
pip install -e ".[azure]"
```

You should then be able to run the example.

**Comment by vsmalladi** [comment]
Where should the output files be stored once the pipeline is done?

**Comment by jakevc** [comment]
In the storage container/prefix specified at runtime by `--default-remote-prefix`

**Comment by vsmalladi** [comment]
Okay ran the update version and am getting an error it can't find the output files:

```
rule bwa_map:
    input: snakemake-tutorial/data/genome.fa, snakemake-tutorial/data/genome.fa.amb, snakemake-tutorial/data/genome.fa.ann, snakemake-tutorial/data/genome.fa.bwt, snakemake-tutorial/data/genome.fa.fai, snakemake-tutorial/data/genome.fa.pac, snakemake-tutorial/data/genome.fa.sa, snakemake-tutorial/data/samples/B.fastq
    output: snakemake-tutorial/mapped_reads/B.bam
    log: snakemake-tutorial/logs/mapped_reads/B.log
    jobid: 5
    reason: Missing output files: snakemake-tutorial/mapped_reads/B.bam
    wildcards: sample=B
    threads: 2
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=<TBD>

{ bwa mem -t 2 snakemake-tutorial/data/genome.fa snakemake-tutorial/data/samples/B.fastq | samtools view -Sb - > snakemake-tutorial/mapped_reads/B.bam; } 2>snakemake-tutorial/logs/mapped_reads/B.log
```

**Comment by jakevc** [comment]
The tutorial expects the input files: `snakemake-tutorial/data/genome.fa, snakemake-tutorial/data/genome.fa.amb, snakemake-tutorial/data/genome.fa.ann, snakemake-tutorial/data/genome.fa.bwt, snakemake-tutorial/data/genome.fa.fai, snakemake-tutorial/data/genome.fa.pac, snakemake-tutorial/data/genome.fa.sa, snakemake-tutorial/data/samples/B.fastq` to exist in the storage account you are using, with `snakemake-tutorial` as the storage container.

**Comment by jakevc** [comment]
Can you post/attach the full output log or error?

**Comment by vsmalladi** [comment]
[2022-12-22T192940.156414.snakemake.log](https://github.com/snakemake/snakemake/files/10292068/2022-12-22T192940.156414.snakemake.log)

**Comment by jakevc** [comment]
Can you confirm that you have all input files where they are suppose to be. `snakemake-tutorial/data/genome.fa, snakemake-tutorial/data/genome.fa.amb, snakemake-tutorial/data/genome.fa.ann, snakemake-tutorial/data/genome.fa.bwt, snakemake-tutorial/data/genome.fa.fai, snakemake-tutorial/data/genome.fa.pac, snakemake-tutorial/data/genome.fa.sa, snakemake-tutorial/data/samples/B.fastq`. Those files should all exist in the storage account you are using. With "snakemake-tutorial" as the container name.

**Comment by jakevc** [comment]
And there should also be `snakemake-tutorial/data/samples/A.fastq, snakemake-tutorial/data/samples/B.fastq, snakemake-tutorial/data/samples/C.fastq`

**Comment by vsmalladi** [comment]
Yep all present

**Comment by vsmalladi** [comment]
okay it looks like it is not pulling in the data from the storage account and I think its trying to get the files from git:

```
Unable to retrieve additional files from git. This is not a git repository.
```

**Comment by vsmalladi** [comment]
Confirmed this works on azure. I think we can merge this in when ready.

**Comment by vsmalladi** [comment]
@jakevc @johanneskoester Any blockers on this PR?

**Comment by jakevc** [comment]
> @jakevc @johanneskoester Any blockers on this PR?

All good on my end

**Comment by johanneskoester** [comment]
Awesome, thanks a lot! Indeed we have been granted credits by MS! I will try to include the tokens into the secrets ASAP.

**Comment by johanneskoester** [comment]
Subsciption to azure is set up, and I have added the credentials as AZ_BATCH_KEY and AZ_STORAGE_KEY in Github Actions. Can you add a test case here?

**Comment by jakevc** [comment]
> Subsciption to azure is set up, and I have added the credentials as AZ_BATCH_KEY and AZ_STORAGE_KEY in Github Actions. Can you add a test case here?

Okay I have added a test case. It requires more secrets, please add the remaining secrets referenced in the workflow:

```
  export AZ_BLOB_ACCOUNT_URL=${{ secrets.AZ_STORAGE_ACCOUNT_URL }}
  export AZ_BLOB_AZ_BLOB_CREDENTIAL=${{ secrets.AZ_STORAGE_KEY }}
  export AZ_BATCH_ACCOUNT_URL=${{ secrets.AZ_BATCH_ACCOUNT_URL }}
  export AZ_BATCH_ACCOUNT_KEY=${{ secrets.AZ_BATCH_KEY }}
  pytest -s -v -x tests/test_azure_batch_executor.py
```

File: tests/test_azure_batch_executor.py
Comment (johanneskoester) [inline tests/test_azure_batch_executor.py:12]: I think the test folder still needs to be added.

File: tests/test_azure_batch_executor.py
Comment (johanneskoester) [inline tests/test_azure_batch_executor.py:16]: Why is a special container image needed?

File: tests/test_azure_batch_executor.py
Comment (johanneskoester) [inline tests/test_azure_batch_executor.py:10]: I am unsure how this should look like and where I will find it. Can you also put this info in the section of the docs? For now, I've set it to `https://snakemakebatch.eastus.batch.azure.com`.

**Review by johanneskoester — CHANGES_REQUESTED** [review]

File: tests/test_azure_batch_executor.py
Comment (vsmalladi) [reply tests/test_azure_batch_executor.py:17]: It shouldn't I think we need to back into base snakemake image

File: tests/test_azure_batch_executor.py
Comment (jakevc) [reply tests/test_azure_batch_executor.py:13]: ug yeah git really wants to ignore that whole directory for some reason

**Comment by jakevc** [comment]
@johanneskoester please make sure the secret AZ_STORAGE_ACCOUNT_URL is formatted as `https://accountname.blob.core.windows.net`

**Comment by johanneskoester** [comment]
> @johanneskoester please make sure the secret AZ_STORAGE_ACCOUNT_URL is formatted as `https://accountname.blob.core.windows.net`

Ok, done!

**Comment by vsmalladi** [comment]
@johanneskoester is this ready to merge or do we need more testing?

**Comment by johanneskoester** [comment]
@jakevc the test fails, can you please have a look?

**Comment by jakevc** [comment]
@johanneskoester  It doesn't look like the test action is picking up the secret values correctly. Can you double check they are present and corresponding to below: 

```
          export AZ_BLOB_PREFIX=${{ secrets.AZ_BLOB_PREFIX }}
          export AZ_BLOB_ACCOUNT_URL=${{ secrets.AZ_STORAGE_ACCOUNT_URL }}
          export AZ_BLOB_CREDENTIAL=${{ secrets.AZ_STORAGE_KEY }}
          export AZ_BATCH_ACCOUNT_URL=${{ secrets.AZ_BATCH_ACCOUNT_URL }}
          export AZ_BATCH_ACCOUNT_KEY=${{ secrets.AZ_BATCH_KEY }}
```

**Comment by jakevc** [comment]
This is the format of the secrets I am using when I run the test locally and it passes:

AZ_BLOB_PREFIX=snaketest
AZ_BLOB_ACCOUNT_URL=https://***********.blob.core.windows.net/
AZ_BATCH_ACCOUNT_URL=https://**********.westus2.batch.azure.com
AZ_BLOB_CREDENTIAL=$SAS
AZ_BATCH_ACCOUNT_KEY=$KEY

**Comment by johanneskoester** [comment]
No clue, I just checked it again. The URL for the storage account seems follow the same pattern as you suggest.

**Comment by johanneskoester** [comment]
Maybe you can add some parsing of those env vars to the remote provider and the executor, which gives a proper error message if something is invalid (without printing any secrets of course).

**Comment by vsmalladi** [comment]
I wonder if there is some special characters that aren't escaped correctly. 

@jakevc can you try dummy variables and print those to stdout and see if there is something odd with the variables.

**Comment by vsmalladi** [comment]
@jakevc  can you do this

          AZ_BLOB_PREFIX: '$${{ secrets.AZ_BLOB_PREFIX }}'
          AZ_BLOB_ACCOUNT_URL: '${{ secrets.AZ_STORAGE_ACCOUNT_URL }}'
          AZ_BLOB_CREDENTIAL: '${{ secrets.AZ_STORAGE_KEY }}'
          AZ_BATCH_ACCOUNT_URL: '${{ secrets.AZ_BATCH_ACCOUNT_URL }}'
          AZ_BATCH_ACCOUNT_KEY: '${{ secrets.AZ_BATCH_KEY }}'

**Comment by jakevc** [comment]
Face palm: of course this is the issue. The funny thing is on my development environment zsh doesn't even let me set the blob credential without quoting it.

**Comment by jakevc** [comment]
or not...

**Comment by jakevc** [comment]
Okay I'm not crazy.. I slightly more specific error and now we are failing with:

```
raise ValueError("Blob Account URL is None or empty string")
[220](https://github.com/snakemake/snakemake/actions/runs/4848828641/jobs/8640274478#step:20:221)
ValueError: Blob Account URL is None or empty string
```

**Comment by johanneskoester** [comment]
Understood the problem now! The issue is that as the PR comes from a fork, it does not have access to secrets.

**Comment by johanneskoester** [comment]
I've merged your changes in a snakemake-owned branch and created a PR for merging into main here: https://github.com/snakemake/snakemake/pull/2246
Let us go on from there. As you are anyway already part of the Snakemake org, you should in the future work directly in our repo in order to ensure that your PRs always have the secrets.

File: snakemake/remote/AzBlob.py
Comment (vsmalladi) [inline snakemake/remote/AzBlob.py:209]: @jakevc this url pattern fails when the sas token is added  should be

`r"^https:\/\/[a-z0-9]+(\.[a-z0-9]+)*\.blob\.core\.windows\.net\/(\?.+)?$"`

**Pull request merged**
